### PR TITLE
[search] Tweaks to search API

### PR DESCRIPTION
**Pull request by ejgallego**
The current `Search.search_*` API returns a large `Pp.std_ppcmds`
object with the formatting of the search results. For large searches this is inefficient as:
- a large list object is generated only to be iterated.
- users must wait for the whole search to complete before seeing any results. 

This PR modifies the API so users can perform per-result processing of
search results. We now emit a feedback message for every result, thus streaming follows.

This PR offers nice improvements to JsCoq, now search
results are printed as they are generated.

### CODE REVIEW

File: dev/doc/changes.txt [hunk 1]
```diff
@@ -1,4 +1,12 @@
 =========================================
+= CHANGES BETWEEN COQ V8.5 AND CQQ V8.6 =
```
Comment (JasonGross) [inline dev/doc/changes.txt]: `CQQ` -> `COQ`

**Comment by JasonGross** [comment]
You should probably close this and open a new PR on trunk instead of on v8.5?

File: toplevel/search.ml [hunk 2]
```diff
@@ -167,8 +165,9 @@ let search_pattern gopt pat mods =
   let iter ref env typ =
     if filter ref env typ then plain_display ans ref env typ
   in
-  let () = generic_search gopt iter in
```
Comment (JasonGross) [inline toplevel/search.ml]: What's the difference between `let () = ... in ...` and `...; ...`?

Comment (ejgallego) [inline toplevel/search.ml]: AFAICT `a; b` is syntactic sugar for `let () = a in b`.

File: dev/doc/changes.txt (hunk 1 above)
Comment (ejgallego) [inline dev/doc/changes.txt]: thx fixed

**Comment by ejgallego** [comment]
@JasonGross given that is not clear what the release cycle will be, this minor changes can be applied to v8.5, which is regularly merged back into trunk AFAICT.

**Comment by ppedrot** [comment]
I think that officially we don't like too much changing the API between pl except for those that are absolutely necessary. In this case I believe it does not matter that much because virtually nobody uses this API yet. Nonetheless, I would like to have opinions of other devs before merging this one.

Ah, and this is not going to be the API of 8.6 anyway, we probably want something more stream-esque.

**Comment by ejgallego** [comment]
Thanks for the comments, I'm doing some more testing and will update this PR and get back soon.

**Comment by ejgallego** [comment]
Update PR, thanks!

**Comment by ejgallego** [comment]
By the way, the only user of `Declaremods.iter_all_segments` is search, so indeed there's some space to have fun here.

**Comment by gmalecha** [comment]
I spent some time this weekend looking into performance hacking on my own branch. I ended up with the code here:

https://github.com/gmalecha/coq/tree/search-interface

The old interface is still there because I didn't want to change everything before I got some more feedback since it is a fairly substantial change. There are three main differences in the code that are basically orthogonal:

First and more interesting, this code inverts the generation of the results; rather than calling a function that calls a [display_function] every time that the search find a result, it constructs a stream of results (see https://github.com/gmalecha/coq/blob/search-interface/lib/cStream.mli). The revised interface includes the following functions:

    val generic_search_stream : int option ->
      (global_reference -> env -> constr -> 'b stream -> 'b stream) ->
      'b stream

This seems a little bit nicer for clients that want to pull results in a lazy style, e.g. if you want only the first 50 results, then you only search until you find 50 things. In the current design, I assume that this would be done by raising an exception from the [display_function].

The other benefit of this interface, in my opinion, is that it is more functional than the previous implementation. The previous implementation constructed lists by writing the results into a [ref] cell. Originally I thought that this would be more of a performance win than it turns out to be, in my tests the code is about the same speed as the old code, though occasionally it is slightly faster. Unfortunately, OCaml's inliner isn't really up to the task of optimizing the interface by itself (this is why the interface uses functions that are ```'a stream -> 'a stream``` rather than just ```'a stream```), so it takes some manual tweaking to get the full performance, I had been hoping that we would be able to use functions such as ```stream_map```, and ```stream_filter``` on the raw results but my current attempts at this introduce a little under 10% increase in running time.

Second, and something that I think is important, is that the ```search_constraint``` type does not use ```string``` anymore. Instead it uses ```regexp``` and ```constr_pattern``` and I provide convenience functions to parse strings into each of those (for ```regexp``` the client just calls ```Str.regexp```). The reason that I think this is important is because accessing this interface from the client side is annoying if you always have to convert you data to strings only to have the interface convert the information back.

The final change, which I'm not sure is necessary, is changing the ```search_constraint``` type to include binary operators. Originally, I thought that this would offer a bit more expressiveness but I'm not exactly sure how it would be exposed to the user so it might not be worthwhile. I'd be happy scrapping this bit completely but if someone has a use-case for it, keeping it in wouldn't require too much additional work.

**Comment by mattam82** [comment]
Looks nice, what do you think @ppedrot, is it streamesque-enough for you?

**Comment by gmalecha** [comment]
Which of the three proposals are you in favor of? @mattam82 @ppedrot

**Comment by mattam82** [comment]
Personally I'm not an expert on the Search interfaces, but I think the move to regexp, constr_pattern and an algebraic search_constraint type goes in the right direction.

**Comment by ejgallego** [comment]
I much support going to streams (or a similar abstractions), at the last WG there was some talk of having an IDE API based on search (so IDEs can query for hypothesis, goals, tactics in scope, etc...) using an unified interface.

My main question here is what the semantics of streams is supposed to be in the presence of the impure STM code. For instance, what should a search do if a new lemma appears in the lifetime of the stream.

At the API level I though of solving this by attaching a STM id to every stream, and requiring the user to close it in order to free + some additional rules, and the plan was to implement that as a simple streaming of a list.

I'm not sure how this should work here thou.

**Comment by gmalecha** [comment]
I think that this is currently in line with what you are saying. The stream interface is purely functional, so it grabs the tree at a particular point (in time) and streams it. So, assuming that there are no imperative updates to the tree (I don't think there are) then I would expect the semantics that you are saying. Personally, I don't think that it makes sense to have the STM update the results of a query as things change.

From the point of view of the interface to IDEs, it seems like the only difference that you would get is that streams would be objects and you could request more objects over time. For example, you would have something like:
```
start_query : .... -> stream_id
pull_query : int -> stream_id -> list results
close_query : stream_id -> unit
```
Personally, this interface seems overly complex and it feels like a simpler interface where you just have asynchronous queries would work better. That interface is easily (and more efficiently) implemented on top of the existing iterator-based interface.

My (personal) motivation for streaming results was so that I could use the search interface from plugins (in particular [my plugin](https://github.com/gmalecha/coq-ltac-iter)) where you don't necessarily want the exhaustive enumeration of things that match.

**Comment by ejgallego** [comment]
Hi Gregory thanks for the comments.

I agree that it does not make sense to update the results as proofs go; however note that in Coq there are many imperative updates to data structures so care has to be taken when streaming them as they may correspond to different "states".

This is why I thought planned to attach queries to Stm `state_id`, to contextualize them. However, I am not sure how garbage collection interacts with this. Let's say in state 101 I create a search stream `a`, which contains 200 results, which the IDE pulls 20.

What is the lifetime of this object? Is is invalidated when the STM goes to a new state? Are the results preserved (thus creating a memory leak)?

**Comment by gmalecha** [comment]
From what I could tell, it seems like the tree of definitions is only imperative at the top level, i.e. there is a top-level reference to a purely functional data structure. Since the stream closes over the current value of the reference (not the reference itself), it should be stable regardless of what the STM does. In theory, it would produce a memory leak if handles were held to it for a long time.

You could manually invalidate certain objects when the system transitions but I still feel like this is a bit overkill. In my mind, coqtop should be as functional as possible (even though it is currently far from it).

**Comment by ejgallego** [comment]
Thanks for the analysis Gregory. IMO the key point here is who is holding the reference to the stream indeed.

When talking about IDEs, it is still not clear to me what the lifetime of streams should be, specially given that we are talking about serialization to outside Ocaml.

**Comment by gmalecha** [comment]
@ejgallego In my opinion the query interface should be something like this:

query : [query specification] -> query_id
kill_query : query_id -> ()

The first allows you to perform a query and the results will be sent back to the user according to Coq's scheduling of the search. In the single-threaded world, that probably means immediately but in the future I could imagine more delayed processing (perhaps in the future it makes sense to add some sort of priority facet to the API). By default, the query is finished and cleaned up by the STM as soon as the results are finished. The kill_query allows the IDE to terminate the query early if it wants to.

An IDE driven interface (where the IDE explicitly asks for the next n results) could also be useful for some things but at the moment, I don't have a good idea of what they would be.

**Comment by ejgallego** [comment]
Hi Gregory, thanks for the feedback! I am trying to write a CEP (https://coq.inria.fr/cocorico/CEP/CEP-0) to propose an extension for the current IDE API, I'll comment more when the doc is ready.

**Comment by ejgallego** [comment]
Up! I got this rebased against master; with this change search results are "streamed" automatically to backends including jsCoq.

I had to open PR #233  due to the retarget to master due to github restrictions.

**Pull request closed without merging**